go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -224,15 +293,41 @@ func runOvnKube(ctx *cli.Context, cancel context.CancelFunc) error {
 	}()
 	var masterWatchFactory *factory.WatchFactory
 	var masterEventRecorder record.EventRecorder
-	if master != "" {
+
+	if runMode.networkControllerManager {
 		var err error
 		// create factory and start the controllers asked for
 		masterWatchFactory, err = factory.NewMasterWatchFactory(ovnClientset.GetMasterClientset())
 		if err != nil {
 			return err
 		}
 		defer masterWatchFactory.Shutdown()
+	}
+
+	if runMode.clusterManager {
+		var clusterManagerWatchFactory *factory.WatchFactory
+		if runMode.networkControllerManager {
+			clusterManagerWatchFactory = masterWatchFactory
+		} else {
+			clusterManagerWatchFactory, err = factory.NewClusterManagerWatchFactory(ovnClientset.GetClusterManagerClientset())
+			if err != nil {
+				return err
+			}
+			defer clusterManagerWatchFactory.Shutdown()
+		}
+
+		cm := clustermanager.NewClusterManager(ovnClientset.GetClusterManagerClientset(), clusterManagerWatchFactory,
+			getClusterManagerIdentity(ctx), wg, util.EventRecorder(ovnClientset.KubeClient))
+		leaderElector := ovnleaderelector.NewOvnLeaderElector("ovn-kubernetes-cluster-manager", getClusterManagerIdentity(ctx), cm)
+		err = leaderElector.Start(ovnClientset.KubeClient, util.EventRecorder(ovnClientset.KubeClient), ctx.Context, cancel,
+			config.ClusterMgrHA, cm.GetMericsProvider(), wg)
```

Comment:
This leader elector utility is nice. Could we keep it internal to `ClusterManager` for now or is there a reason not to?

Reply:
Ok.  I'll keep it internal to cluster manager for now.  My plan is to use this the following way

  1.  For "ovnkube --init-master",  create only one leader election using this utility with the election name  "ovn-kubernetes-master" and once the ovn leader elector instance wins the election, it will start both cluster manager and network controller manager.
  2.  For "ovnkube --init-cluster-manager",  create only one leader election using this utility with the election name  "ovn-kubernetes-cluster-manager" and once the ovn leader elector instance wins the election, it will start cluster manager.
  3.  For "ovnkube --init-network-controller-manager",  create only one leader election using this utility with the election name  "ovn-kubernetes-network-controller-manager" and once the ovn leader elector instance wins the election, it will start network controller manager.
  

With this present PR,  we create 2 leader elections (ovn-kubernetes-master and ovn-kubernetes-cluster-manager) which I think isn't that great.

wdyt about the proposed approach ?  I don't want to disrupt too many things and hence didn't include in this PR.  I also need to work on it and test it out.  If the approach sounds good,  we can have a follow up patch and move the leader elector out of the clsuter-manager.

Reply:
These plans make sense.

But I was not clear enough. I did not mean to say to move the leader elector utility to the cluster manager package. It is fine that it is exported through its own package. I meant to keep the leader elector instance as a private property of the cluster manager until those plans above materialize. Otherwise it looks a bit inconsistent to not be using that leader elector for the other components as well.

Reply:
I've moved  the election code back to ClusterManager.Start() for now.  

ClusterManager has now only to public functions - Start and Stop.  Test code also calls Start() and Stop().   If the test calls clustermanager.Start() with nil event recorder,  the election works fine with the fake client.